### PR TITLE
fix(bigtable): Throw on unsuccessful set/delete operations

### PR DESCRIPTION
**Pull request by tkaemming**
The [`commit` method on direct row operations](https://googleapis.dev/python/bigtable/latest/row.html#google.cloud.bigtable.row.DirectRow.commit) will only throw in the case when a request is determined to be inherently unable to succeed before it performed. Operations that fail on the server side are returned as [`Status` values](https://github.com/googleapis/googleapis/blob/master/google/rpc/status.proto) which must be handled by the caller.

Internally, [`commit` calls `Table.mutate_rows`](https://github.com/googleapis/python-bigtable/blob/cf3b9a1820780863911f4741fca4e9936cb79670/google/cloud/bigtable/row.py#L466) which more clearly highlights this behavior in the API documentation: https://googleapis.dev/python/bigtable/latest/table.html#google.cloud.bigtable.table.Table.mutate_rows

These operations are [retried internally with a fairly comprehensive retry policy](https://github.com/googleapis/python-bigtable/blob/cf3b9a1820780863911f4741fca4e9936cb79670/google/cloud/bigtable/table.py#L60-L66) but given the importance of this component, we should be extra mindful of failure handling here.

I have manually verified the client behavior is as documented by using the `cbtemulator` backend and attempting to write a value against a column family that does not exist (this is the easiest type of error I could think of to reproduce, I am not sure what others may exist in an actual environment):

```
Traceback (most recent call last)
<ipython-input-1-e5166fac7e8c> in <module>
---> 11 ns.set("key", "value")

~/workspaces/sentry/sentry/src/sentry/nodestore/base.py in set(self, id, data, ttl)
    202         >>> nodestore.set('key1', {'foo': 'bar'})
    203         """
--> 204         return self.set_subkeys(id, {None: data}, ttl=ttl)
    205 
    206     def set_subkeys(self, id, data, ttl=None):

~/workspaces/sentry/sentry/src/sentry/nodestore/base.py in set_subkeys(self, id, data, ttl)
    220         cache_item = data.get(None)
    221         bytes_data = self._encode(data)
--> 222         self._set_bytes(id, bytes_data, ttl=ttl)
    223         # set cache only after encoding and write to nodestore has succeeded
    224         self._set_cache_item(id, cache_item)

~/workspaces/sentry/sentry/src/sentry/nodestore/bigtable/backend.py in _set_bytes(self, id, data, ttl)
    180         status = row.commit()
    181         if status.code != 0:
--> 182             raise BigtableError(status.code, status.message)
    183 
    184     def encode_row(self, id, data, ttl=None):

BigtableError: (13, 'unknown family "x"')
```

I am actively working on extracting the key/value storage parts of this class so that it can be reused in other contexts and will be replacing the existing tests with tests that actually use a running emulator and exercise the real bigtable client rather than the mocked backend, so I did the minimum to get the existing tests to pass as written.

### CODE REVIEW

**Comment by tkaemming** [comment]
@untitaker, @wedamija, @lynnagara: I added you as reviewers because GitHub said you were recently in this file and the original author/domain owner no longer works here. If you think somebody else would be a better fit feel to add them and remove yourself.

**Comment by tkaemming** [comment]
I'd appreciate a review on this given our outage this morning involving these two methods calls.

**Comment by lynnagara** [comment]
This looks ok to me. Though I'm slightly worried about the volume of exceptions we'd be sending into Sentry if bigtable writes started failing again

**Comment by tkaemming** [comment]
> I'm slightly worried about the volume of exceptions we'd be sending into Sentry if bigtable writes started failing again

I understand (and to some extent share) the concern, but I also don't see how this is different from what I think the assumed expected behavior was (or at least what I hope it was, I don't think it was ever meant to potentially fail silently), or how this is different from any other external systems within the processing pipeline that might also fail. If the choices are "silently fail and lose and/or corrupt data" or "crash and allow the caller to preserve data integrity or make data loss obvious" I don't think it's much of a decision, even if that crashing happens to occur at an otherwise inconvenient time.

**Comment by untitaker** [comment]
>will be replacing the existing tests with tests that actually use a running emulator and exercise the real bigtable client rather than the mocked backend

There are two parametrizations, one in [`test_common`](https://github.com/getsentry/sentry/blob/master/tests/sentry/nodestore/test_common.py), one in the file you are touching here, that are currently skipped that would basically that. I added the mocked backend because they were the only tests and as such nothing ran in CI. I'd like to keep both tbh and just stop skipping the "real client tests"

File: src/sentry/nodestore/bigtable/backend.py [hunk 1]
```diff
@@ -231,7 +238,11 @@ def delete(self, id):
 
         row = self.connection.row(id)
         row.delete()
-        row.commit()
+
+        status = row.commit()
+        if status.code != 0:
+            raise BigtableError(status.code, status.message)
```
Comment (untitaker) [inline src/sentry/nodestore/bigtable/backend.py:244]: any way we can get that impld for delete_multi too?

**Review by untitaker — APPROVED** [review]

File: src/sentry/nodestore/bigtable/backend.py (hunk 1 above)
Comment (tkaemming) [reply src/sentry/nodestore/bigtable/backend.py:244]: Oops, yeah. That was an oversight on my part, will include there as well. Thanks for the reminder.

**Review by tkaemming — COMMENTED** [review]
> I added the mocked backend because they were the only tests and as such nothing ran in CI. I'd like to keep both tbh and just stop skipping the "real client tests"

That makes sense, will keep this in mind as I move forward and also keep you in the loop with these changes.

File: src/sentry/nodestore/bigtable/backend.py (hunk 1 above)
Comment (tkaemming) [reply src/sentry/nodestore/bigtable/backend.py:244]: Added in e9d96fc4612262b5392c2a2762f6646ea372dde3, thanks again for calling that out.

File: src/sentry/nodestore/bigtable/backend.py [hunk 2]
```diff
@@ -248,8 +259,18 @@ def delete_multi(self, id_list):
...
-        self._delete_cache_items(id_list)
+        deleted_ids = []
+        errors = []
+        for id, status in zip(id_list, self.connection.mutate_rows(rows)):
+            if status.code == 0:
+                deleted_ids.append(id)
+            else:
+                errors.append(BigtableError(status.code, status.message))
+
+        self._delete_cache_items(deleted_ids)
```
Comment (tkaemming) [inline src/sentry/nodestore/bigtable/backend.py:270]: I have a suspicion there are circumstances where a delete might succeed be represented as a failure (e.g. network timeouts or failures to meet requested consistency levels for replication), whether or not this is the right approach to dealing with those kinds of failures will probably need to be re-evaluated at some point (same goes for single row deletion or any row mutation, I guess.)

**Pull request merged**